scripts/build_test.sh

```diff
@@ -21,38 +21,187 @@ if [[ -n "${NO_RACE:-}" ]]; then
-# This is useful for flaky tests
+# MAX_RUNS bounds the attempts to retry individual tests before giving up
 MAX_RUNS=4
-for ((i = 1; i <= MAX_RUNS; i++));
-do
-    # shellcheck disable=SC2046
-    go test -shuffle=on ${race:-} -timeout="${TIMEOUT:-600s}" -coverprofile=coverage.out -covermode=atomic "$@" $(go list ./... | grep -v github.com/ava-labs/coreth/tests) | tee test.out || command_status=$?
-
-    # If the test passed, exit
-    if [[ ${command_status:-0} == 0 ]]; then
-        rm test.out
-        exit 0
-    else 
-        unset command_status # Clear the error code for the next run
+
+# Function to extract failed test names from output
+extract_failed_tests() {
+    local output_file="$1"
+    # Extract test failures and panics
+    (grep "^--- FAIL" "$output_file" | awk '{print $3}' || grep -E '^\s+Test.+ \(' "$output_file" | awk '{print $1}') |
+    sort -u
+}
+
+# Function to check if a test is a known flake
+is_known_flake() {
+    local test_name="$1"
+    grep -q "^${test_name}$" ./scripts/known_flakes.txt
+}
+
+# Function to check for panics in output
+detect_panics() {
+    local output_file="$1"
+    grep -q "panic:" "$output_file" || grep -q "fatal error:" "$output_file"
+}
+
+# Function to get all tests in a package
+get_package_tests() {
+    local package="$1"
+    go test -list=".*" "$package" 2>/dev/null | grep "^Test" || true
+}
+
+# Function to get tests that didn't run due to panic
+get_missing_tests_after_panic() {
+    local output_file="$1"
+    local package="$2"
+    local failed_tests="$3"
+    
+    # Get all tests in the package
+    local all_tests
+    all_tests=$(get_package_tests "$package")
+    
+    # Find tests that didn't run (not in failed_tests and not in successful output)
+    local missing_tests=""
+    for test in $all_tests; do
+        # Check if test is in failed_tests
+        if echo "$failed_tests" | grep -q "$test"; then
+            continue
+        fi
+        
+        # Check if test passed (appears in output with PASS)
+        if grep -q "PASS.*$test" "$output_file"; then
+            continue
+        fi
+        
+        # Test didn't run
+        missing_tests="$missing_tests $test"
+    done
+    
+    echo "$missing_tests"
+}
+
+# Function to run specific tests and return status
+run_specific_test() {
+    local test_name="$1"
+    local output_file="$2"
+    
+    # Extract package name from test name (assuming format: PackageName.TestName)
+    local package
+    package="${test_name%.*}"
+    
+    # Run the specific test
+    go test -shuffle=on ${race:-} -timeout="${TIMEOUT:-600s}" -run "^${test_name}$" "$package" 2>&1 | tee "$output_file"
+    return "${PIPESTATUS[0]}"
+}
+
+# Initial test run
+echo "Running initial test suite..."
+# shellcheck disable=SC2046
+go test -shuffle=on ${race:-} -timeout="${TIMEOUT:-600s}" -coverprofile=coverage.out -covermode=atomic "$@" $(go list ./... | grep -v github.com/ava-labs/coreth/tests) | tee test.out || command_status=$?
```

Comment:
Not sure if it's possible to write this line so it both 
A. works
B. passes lint

Reply:
Interesting! That's definitely something that is worth bringing over to the evm repositories in my opinion.